File: mobile/src/main/java/com/crow/stand_reminder/tool/DatabaseTools.java
package com.crow.stand_reminder.tool;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;

import androidx.room.Room;

import com.crow.stand_reminder.data.Database;
import com.crow.stand_reminder.data.SensorValue;

public class DatabaseTools
{
	private Database database;

	private Context context;

	public DatabaseTools(Context context)
	{
		this.context = context;

		database = Room.databaseBuilder(context, Database.class, "storage.db").build();
	}

	@Override
	protected void finalize() throws Throwable
	{
		super.finalize();
		database.close();
	}

	public void drop()
	{
		database.clearAllTables();
	}

	public void saveValue(float value)
	{
		database.values().insertAll(new SensorValue(value));
	}

	public void saveValue()
	{
		SensorTools sensorTools = new SensorTools(context);

		sensorTools.getSensorManager().registerListener(new SensorEventListener()
		{
			@Override
			public void onSensorChanged(SensorEvent event)
			{
				// We only want one value
				sensorTools.getSensorManager().unregisterListener(this);
				// It's 1 on mobile, 0 on watch
				saveValue(event.values[1]);
			}

			@Override
			public void onAccuracyChanged(Sensor sensor, int accuracy)
			{
			}
		}, sensorTools.getSensor(), SensorManager.SENSOR_DELAY_NORMAL);
	}
}

File: mobile/src/main/java/com/crow/stand_reminder/tool/DatabaseTools.kt
package com.crow.stand_reminder.tool

import android.content.Context
import androidx.room.Room
import com.crow.stand_reminder.data.AppDatabase
import com.crow.stand_reminder.data.CompletedHour
import java.io.Closeable
import java.util.*

class DatabaseTools(context: Context) : Closeable
{
	private val database: AppDatabase =
		Room.databaseBuilder(context, AppDatabase::class.java, "storage.db").build()

	override fun close() =
		database.close()

	private fun completed() =
		database.completed().values

	private fun compareCalendar(date1: Calendar, date2: Calendar, field: Int) =
		date1.get(field) == date2.get(field)

	fun isCompleted(date: Calendar) =
		completed().any { value ->
			compareCalendar(value.date, date, Calendar.YEAR) &&
				compareCalendar(value.date, date, Calendar.MONTH) &&
				compareCalendar(value.date, date, Calendar.DAY_OF_MONTH) &&
				compareCalendar(value.date, date, Calendar.HOUR_OF_DAY)
		}

	/**
	 * Sets minute, second and millisecond to 0
	 */
	private fun Calendar.resetMinutes()
	{
		(this).apply {
			set(Calendar.MINUTE, 0)
			set(Calendar.SECOND, 0)
			set(Calendar.MILLISECOND, 0)
		}
	}

	private fun Calendar.toResetMinutes(): Calendar =
		this.apply {
			resetMinutes()
		}

	private fun Calendar.toResetHours(): Calendar =
		this.apply {
			resetMinutes()
			set(Calendar.HOUR_OF_DAY, 0)
		}

	fun addCompleted(completed: CompletedHour)
	{
		// Ignore minutes and seconds
		completed.date.resetMinutes()
		// Add it to the database
		database.completed().insertAll(completed)
	}

	fun getCompletedToday(): List<CompletedHour> =
		Calendar.getInstance().toResetHours().let {
			return completed().filter { c ->
				c.date.toResetHours() == it
			}
		}

	fun drop() =
		database.clearAllTables()
}